### ISSUE TITLE
NetStream rtmp live video stream not working with Android since Adobe Air 33.1 - Paying User For License - Please respond

### ISSUE THREAD

**Issue by apofis1969** [Gamua/Adobe-Runtime-Support#330]
Problem Description
The NetStream feature does not work with Android since Adobe Air 33.1
If Android is the Receiver of a rtmp live video stream the sound and video transmission is broken. Tried a lot of workarounds, different rendering modes, but no chance. All emails I sent to Harman stay without any response and this problem is still not fixed.
Maybe someone here has an idea or solution.
The last working version of the Air SDK is 33.0.2.330
We bought the License to reject the Harman Splash Screen. The company is CHCNET in Austria.
So please send us an answer when this problem will be fixed.
Thank you!

https://tracker.adobe.com/#/view/AIR-4198908

**Comment by ajwfrost** [comment]
Hi - we have some restrictions at the moment with a/v due to some of the licensing issues i.e. we didn't get the full source code from Adobe, they aren't able to pass us some of the third party a/v code. 

But if you're saying it worked in 33.0.2.330 on Android, and now doesn't work, then it may be a different problem here! Particularly given your comment on the Adobe Tracker:

> Rendering mode GPU, audio stream is broken, video stream is perfect.
> All other rendering modes have a broken video stream, but a perfect audio quality.

I can ask one of the Android guys to look into this and to see what might have changed in the code, I don't suppose you can provide a demo app that we can use to test it?

thanks

**Comment by apofis1969** [comment]
Hi, if you send me a link where I can send the apk and the instructions to test I will send you the apk build with the new Air SDK.

**Comment by ajwfrost** [comment]
Thanks @apofis1969  - please use the below link:
https://transfer.harman.com/requests/Cv02JuPIF1y1RE5G7qjXbC

**Comment by Marcophono2** [comment]
Same problem here. RTMP audio live streaming to my app is working but with increasing delays and pulling down the whole app within seconds until the screen is frozen. 2 mio active app users. I'm becoming nervous a bit now due to Andrew's post that there is a problem with missing source codes from Adobe..

**Comment by apofis1969** [comment]
It is a mess. And they charge you to remove the splash screen for a sdk which is not working. It is a shame.

**Comment by apofis1969** [comment]
I was trying to send the apk but the link has expired:
https://transfer.harman.com/requests/Cv02JuPIF1y1RE5G7qjXbC

**Comment by Marcophono2** [comment]
> It is a mess. And they charge you to remove the splash screen for a sdk which is not working. It is a shame.

I agree. Aside from the fact that Harman ist dimensions more trustful and engaged than Adobe ever was, I indeed think that buying a licence as the only workaround for a "SDK not working at all"-bug is a no-go. Personally I would accept it just to give a (very) little support for the great work they offer, but finally it did not help me because there is another bug in the live streaming, in my case audio live streaming, which appeared first time in the very first version of SDK 33. So the 199 USD licence did not help me to build a working app.

Best
Marc

**Comment by ajwfrost** [comment]
Hi

Updated link for this is:
https://transfer.harman.com/requests/dfHCqlSzlQPTi9zCE97HVy

If you can provide an APK built using "apk-debug" then it means we can start attaching our debug versions of the runtime to see what's happening within the a/v handling code.

In terms of the lack of some source code from Adobe, these were affecting the desktop platforms - it's basically the software decoders for AAC and AVC that Adobe were licensing but that weren't available to us as a sublicense. On mobile platforms these were already using OS-provided (and generally hardware accelerated) APIs. We're looking to rework a lot of the a/v code though as it's a very complex piece of software built up over a long time with many different options, making it a big problem to maintain.

But the issue on Android appears to have been introduced by us more recently - I'd looked through our changes, there was a minor fix we added to correct a seek(0) problem for videos but I can't see how that would affect RTMP in this way. We'll aim to get this fixed, we do see this as an important aspect of the runtime and are working with Veriskope as well to ensure that the whole AMS + AIR set-up continues to work (once we've fixed this current mess!)

thanks

**Comment by Marcophono2** [comment]
@ajwfrost That is wonderful, thanks a lot, Andrew! I'll attach it later today.

Best 
Marc

**Comment by Marcophono2** [comment]
@ajwfrost Hi Andrew! I uploaded my app to https://transfer.harman.com/requests/dfHCqlSzlQPTi9zCE97HVy a week ago. Just curious what the next steps are. Will I be informed about test results? As far as I can see you still didn't make a test. But I can be wrong here if you did a test call in one of the short time frames when my mobile was switched off.

Best regards
Marc

**Comment by ajwfrost** [comment]
Hi @Marcophono2  - sorry for the delay, we'd had a few other issues that the team had been looking at but are now getting on to this one. We can confirm that we're able to load your application on top of our build of the core library so will be able to use this for debugging..

Can I check whether we can call any phone number in order to test this? or if it needs to be yours, what timezone are you in and what hours would be acceptable - we don't want to wake you up! We're likely to need to do quite a lot of brief calls to try to narrow down on what's going on within the runtime.. it could get quite annoying, unless you're able to automatically divert these ones to voicemail perhaps!

thanks

**Comment by ajwfrost** [comment]
FYI - making an assumption on your timezone given your phone number! -  a single call and the logs are showing that we're spending increasingly more time in a callback from the UI thread where we're processing RTMP data - although all the processing is actually done on a different thread, and this background thread looks like it's running too much - perhaps waking up too much. Almost reminds me of an issue we had very early on with the ARMv8 code regarding mutex/thread locking - just out of interest, have you tried an ARMv7 build of your app to see if this has the same problem?

We will delve into it, add more logging in the appropriate places, and have another go...

thanks

**Comment by Marcophono2** [comment]
@ajwfrost I love your team's engagement! Just wonderful! 
Yes, feel free to call my phone number for testing purposes whenever you want. Youj cannot disturb me. If you want to call other phone numbers for testing just send them to my email address so I can put them onto the white list. 
I will do the ARMv7 build test right now.

Best regards
Marc

**Comment by Marcophono2** [comment]
@ajwfrost I was rather sure that there is no difference in an ARMv7 exported apk file but wanted to prove it again what I did now: Same problem. You asked for more logging. I can send you the full souce to help finding this issue if you want. There are too many possible events occuring between the RTMFP streaming and the UI processing where interferrings may happen. Right now I have no idea where to place additional trace logging commands.  

Best regards
Marc

**Comment by ajwfrost** [comment]
@Marcophono2 okay thanks for that.. shame as that would really have helped narrow it down!

The comment re adding more logging was for ourselves i.e. within the various audio threads that are going on, we need to see a little more why they're so busy.

It is starting to look like the problems are just caused by one of the audio threads spinning out of control, it's meant to throw in delays to sleep when it's caught up with processing data but we don't see that happening - or rather, there's half a second of just this thread doing things and then it releases and the main UI thread can continue.

thanks

**Comment by ajwfrost** [comment]
Hi @Marcophono2  - we've looked into this more, and the issue seems to be caused by the AIR use of Android's audio functionality. The problem is happening when we have a very low bitrate audio stream - in this case, some 8kHz mono audio (basically, old-style phone quality). Normally we fill up the Android buffers in 4kb chunks, but with this audio stream we just get 320 bytes at a time and it doesn't seem like Android copes with this...

We're able to ensure that the UI is responsive by changing the write to an asynchronous one, but we also need to then yield/sleep that feeder thread as otherwise we end up getting ahead of the Android buffer capabilities and it starts dropping packets. But now we're facing an issue where the audio is getting jittery i.e. we can get a smooth UI responsiveness, but at the cost of audio quality. If you're able to listen to your voicemails, you'll be able to hear this!

This whole area is one that we need to revisit and will be covered by our new multimedia framework, so I would hope in the not-so-distant future that we will have a fix for this sort of thing. In the meantime, if you're able to change the audio formats so that it's a higher bitrate format then bizarrely I think this might help.

If you prefer the idea of a responsive UI but audio which sounds like it's got little clicks going on the whole time, then please let me know.. apologies we've not been able to get a suitable solution in the short term, I've been getting the impression that Android sound output has always been a bit of a problem!

thanks

**Comment by Marcophono2** [comment]
@ajwfrost Andrew, I could not awaite more than this short-term intensive investigation on low level with this issue identification. Thank you very much for this!!! Indeed an interesting technical background you uncovered. I will try to improve the audio quality on server side in 2 hours when my service is in sleeping mode anyway. I will update you about the result. 
Due to the fact that this issue came up in AIR 33 obviously Harman changed something at the audio processing end. Did you change so much here so that the reason for my issue cannot be located exactlly? I am just curious and interested in the technics behind it. Or did the whole signal/thread processing engine changed?

Best regards
Marc

**Comment by ajwfrost** [comment]
@Marcophono2 I had missed the fact that it was better earlier on... let me see if we can switch this over to 33.0.2 and see what happens with that...

There was a change in this code though, a couple of things where we had to add thread safety and the like. 33.0.2.330 was 28th November and after that, there were two changes to the file that handles the Android audio tracks:
a) adding null-checks around the buffers and ensuring we clean up the track on 'release'
b) changing how our thread loop works with extra critical sections

I can't see how these changes would affect things .. but if I can get your app to run with a .330 runtime, I'll see whether this helps. (Just looking further up this thread actually - it was someone else who said it was working with 33.0.2.330.. do you know whether you found the same though?)

**Comment by Marcophono2** [comment]
@ajwfrost No, I did not read a thread where someone pointed to 33.0.2.330 retated to an audio issue. I would have tried older AIR versions but was not able to find a download link. So large HHDs and SSDs meanwhile with TBs of empty space and I delete files from time to time. Crazy. :) But that certain version I never had. May be you can send me a download link?

**Comment by ajwfrost** [comment]
@Marcophono2  I can easily send you a link - Windows or Mac?  (at least, assuming our file transfer system likes me today..)
I've just tried our trick of replacing just the libCore.so file with one built from 33.0.2.330, but the app doesn't start - likely some differences in JNI which will prevent this from happening. So if you can try it with 330 and let us know if it works, we can then probably stick with our 33.1.1.98 version and start rolling back updates just in the relevant files...

I have a bit more hope now :-)

thanks for the support!

**Comment by Marcophono2** [comment]
@ajwfrost Wonderful! For Windows please.

**Comment by Marcophono2** [comment]
@ajwfrost Sorry, I saw right now that I also tried an older version than the 33.0.2.330. I tried it a few weeks ago with 33.0.2.288. So I think it would be a big surprise if 33.0.2.330 would work but let's try it! :-))

**Comment by Marcophono2** [comment]
@ajwfrost @marchbold  May be, just not to completely exclude that option, together with AIR 33 I had to update all my ANEs and also all Android helper ANEs from distriqt. Do you think the issue could have come up in the ensemble playing between Android, AIR33 and the updated ANEs? Unfortunatelly I cannot test this by changing out it step by step. AIR33 needs the 64bit updated ANEs and that updated ANEs need the updated Android helper ANEs. Also the other way round, using AIR32 with the updated ANEs, cannot work.

**Comment by ajwfrost** [comment]
Link for AIR 33.0.2.330: https://transfer.harman.com/message/ciIE93Ik9n9oT5lQpnqpfQ

@Marcophono2 no from what I can see this is all completely independent of any ANEs that there might be. The sound stuff that we've had to change was something that Adobe had only just added themselves, from what I can see of their own forums and bug reports. If you've got it working with AIR 32, in 32-bit mode, then that would be good; if it then worked also in one of our very first releases, then it might indicate that some of the other changes we had made around this sound area had been the culprit. So we can definitely still focus in on this glue to the Android audio track class - probably we can put back in the original format of the file and see if that helps (although then we'll get stability complaints... but at least we can try to look at the refactor again...)

thanks

**Comment by Marcophono2** [comment]
@ajwfrost @marchbold Me again! I am really sorry, indeed it WORKS with 33.0.2.288! I only can imagine that I had another issue with that version which I fixed meanwhile and mixmatched that issues. So, yes, probably like the OP wrote I also can use up to version 33.0.2.330. That brings me a lot of time because I can roll out an app update basing on one of the older AIR 33 versions. Sorry for having been a stress maker! Yes sure, later this issue will become serious for me when there is another "must" for an AIR update. But so far - I am happy! And yes, Andrew, you solved it for me because you pointed to that older version which I did not had on my screen anymore! And I can quit thinking about activating a video stream (though only needing audio) to increase the bandwidth. :)

**Comment by Marcophono2** [comment]
@ajwfrost You can count on me whenever you need someone to test adjusted AIR versions to find out if that issue is fixed again. I really feel big thanks for your wonderful engagement and would be happy if I can give a part back. However.

**Comment by ajwfrost** [comment]
@Marcophono2  thanks - but can I clarify please, as I just sent a link for 33.0.2.330 .. did it work for you with this one, or with 33.0.2.288 which you may have had still lying around? We will still look at these differences to see whether we can get it working with the latest version ...

cheers

**Comment by Marcophono2** [comment]
@ajwfrost It worked with .288. I'll do the test with .330 now. The OP wrote .330 worked for him, okay. On the other side there were indeed some audio related adjustments implemented in .330.... We will see in a few minutes!

**Comment by Marcophono2** [comment]
.330 also working!

**Comment by Marcophono2** [comment]
@ajwfrost 

```
3.3.3 AIR 33.1.0.16
AIR-300: Preventing hang when switching Wi-Fi connection when RTMFP is being used in a Worker (Gamua-96)
```

Doesn't this looks like being a candidate?
Maybe in dependency with
`AIR-296: Adding further fix to protect from crash in audio code`

**Comment by apofis1969** [comment]
It would be time that there is a working version of Adobe Air for Android. 33.0.2.330 creates a lot of warnings at Google Play in the meanwhile with Android X
We are loosing a lot of clients. Difficult situation.

**Comment by apofis1969** [comment]
@ajwfrost Unfortunately with the new SDK is the same problem. All works perfect, but the video stream is still broken. Please resolve this. We loose so many clients....

**Comment by naka-ok** [comment]
Hello, I have a similer problem about live audio.
I made my application compatible with 64bit, but I am suffering from the problem 
that NetStream.play() causes ANR on Android devices.
No error message is displayed and my application freezes.

When the NetStream consists of video only, NetStream.play() works correctly.
But If the NetStream contains "live audio", it freezes.

Now I'm using AIR SDK version 33.1.1.190.
Will it work if I use version 33.0.2.330 or 33.0.2.288 ?
If so, can you let me download the older versions ?

Thanks

**Comment by Jasenvm** [comment]
Hi All

Having same issue as above. 33.0.2.339 works fine 
However, 33.1.1.190 is not working unless I user`var __microPhone.codec = SoundCodec.PCMU` when publishing NetStream.But still the video is very (very) laggy.

I vote this bug up +1

Hope this is useful.
Great work guys!

**Comment by apofis1969** [comment]
@ajwfrost when will you finally fix the streaming problem. Please we get in huge trouble already. 33.0.2.330 is not very compatible anymore with all the new ANEs and Google Play. Also it is not possible to build the new aab bundle. I am a paying user and can not use the newest version Adobe Air for Android. We are loosing a lot of money already!!

**Comment by ajwfrost** [comment]
@apofis1969  Trying to get some more focus on this but I'm wondering if there are multiple different problems being reported in this thread as it's a little difficult seeing whether these are all the same... we'd worked on one issue which was an audio quality problem but from what I can see:

> If Android is the Receiver of a rtmp live video stream the sound and video transmission is broken. Tried a lot of workarounds, different rendering modes, but no chance

> RTMP audio live streaming to my app is working but with increasing delays and pulling down the whole app within seconds until the screen is frozen

> All works perfect, but the video stream is still broken [should this be 'audio works perfect'?]

> When the NetStream consists of video only, NetStream.play() works correctly. But If the NetStream contains "live audio", it freezes.

> 33.1.1.190 is not working unless I use var __microPhone.codec = SoundCodec.PCMU when publishing NetStream


We can look at some of these but to ensure we focus on exactly what goes wrong for you, would it be possible for you to create a build of your application using the latest working SDK (33.0.2.338?) and built with 'apk-debug', and attach this using the below link and any instructions on how to reproduce if it's not obvious?
https://eu-transfer.harman.com/requests/LgyQMyEgCVdDnFoQtnoIBQ

thanks

**Comment by Marcophono2** [comment]
@ajwfrost Andrew, as you may remember I have the same problem. My app works with the SDK up to 33.0.2.330. Later versions last in increasing audio delays with a full freeze after some seconds. (I have only live audio streaming, no video). I already uploaded an apk in debug mode but with a later, not working SDK. As you asked apofis1969 to build an apk with the last WORKING sdk, I would like to now if I also should upload you a debug version of my app coded with 33.0.2.330?!

Best regards
Marc

**Comment by ajwfrost** [comment]
@Marcophono2  thanks - although I'm not sure that this is the same problem, you're describing an increasing audio delay where the UI was getting frozen as the audio was prioritised over this, and then everything was getting jittery. That sounds to me like a different problem from these other ones that are reported!

**Comment by apofis1969** [comment]
@ajwfrost Hi Andrew, just sent you the files. One with SDK 33.0.2.338 the other with 33.1.1.217
Hope you can find the bug.

**Comment by apofis1969** [comment]
@ajwfrost Hi, I sent you new apks with my google drive link. If you want that I upload here, please send me a new link.
Thanks, Johannes

**Comment by ajwfrost** [comment]
@apofis1969 @Marcophono2  so I was wrong about this being a different issue .. from the video that Johannes shared and from the test APKs, I could see that there was (at least initially) a slow-down in the frame updating when the a/v stream was live, which is what we saw before and tried to tweak the threading/packet handling to improve.

Having reverted one of the more recent changes in the audio handling code, we're back to it working again .. however, I believe that change was made in response to either an ANR (complete thread hang) or a crash (due to a synchronisation problem in the audio threads)... so we may need to just see what alternative approaches we have (and why those changes had this effect!)

thanks

**Comment by apofis1969** [comment]
@ajwfrost Dear Andrew, I want to say thank you to resolve the streaming! Works great with 33.1.1.259! Thanks a lot!

**Comment by Marcophono2** [comment]
@ajwfrost Oh damn, I see that a few days ago a new release came out. As @apofis1969 says it seems to have solved that streaming issue, that is wonderful and I also thank you very much! Meanwhile I became a bit nervous when Google removed an app from Google Play without any warning pointing to a policy violation. It came out that a certain german word from the app description, existing since years, was miss-translated to "fuck". No idea why and how because even the Google translator translates it correctly to "kidding". Finally the app was replaced to the Play Store but that made me think about the dependency on an obviously bad managed control instance on Google side where the step to ask a native german speaking stuff member is skipped and an app is simply removed. To be honest that really shocked me. That was just a half step away from the need to upload a new coded apk with a higher version number and related to the streaming issue that would half been the end for my app and also for my business. So: Many thanks, Andrew!

**Comment by ajwfrost** [comment]
@Marcophono2 @apofis1969  apologies for the delays ..  and thanks for helping us debug things with your apps. We believe we managed to adjust our earlier fix which needed the thread-safety, whilst not having the impact that we'd caused here on the streaming, but as a slight caveat, please let us know if you see an increase in crash reports or ANRs...

thanks

**Comment by apofis1969** [comment]
@ajwfrost I will monitor this. Today I made the update for Android. I realised that the enhanced Microphone is working better with Android than with Apple. maybe you can have a look at this too. The other thing are rotation problems with iOS Phones in Landscape. Moving the app first to the right, the rotation works perfect in landscape. But moving the phone first to the left and then to the right the rotation does not work properly. The layout changes. Appears like a rendering problem. The app does not move to the right smoothly, flips.

**Comment by ajwfrost** [comment]
@apofis1969 thanks - for that iOS rotation thing we are looking into something similar but I don't think we've got a bug for the exact behaviour you're describing .. are you okay to open a new one for that?

For this one .. if you're happy with the fix in 33.1.1.259, please feel free to close the issue :-)

thanks
